### PR TITLE
[API][SIMS-#644]-Receive_MSFAA_File

### PR DESCRIPTION
**Pull request by guru-aot**
- [x] Using similar process to CRA, create fixed fixed file to READ MSFAA File on zoneB SFTP server
- [x] Populate columns in msfaa_numbers table  date_signed | service_provider_received_date

### CODE REVIEW

File: sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts [hunk 1]
```diff
@@ -51,7 +52,155 @@ export interface MSFAARequestFileLine {
...
+
+  /**
+   * Date that the borrower indicated that the MSFAA was Signed
+   */
+  public get borrowerSignedDate(): Date {
+    return dayjs(this.line.substr(23, 8), DATE_FORMAT).toDate();
+  }
+
+  /**
+   * Date MSFAA was received by/resolved from Canada Post/Kiosk
```
Comment (ann-aot) [inline sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts:169]: from `Canada Post/Kiosk`? can you explain this?

File: sources/packages/api/src/msfaa-integration/msfaa-response.service.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      result.errorsSummary.push(logMessage);
+    }
+
+    return result;
+  }
+
+  /**
+   *
+   * @param statusRecord
+   * @param receivedDate
```
Comment (ann-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-response.service.ts:88]: I see here only one parameter `statusRecord`.  pls update the comment

**Review by ann-aot — COMMENTED** [review]
LGTM 👍  Added some minor comments

File: sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts (hunk 1 above)
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts:169]: These are comments from the documentation

image

File: devops/Makefile [hunk 3]
```diff
@@ -238,6 +238,7 @@ init-api:
 	test -n $(API_INIT_ATBC_PASSWORD)
 	test -n $(API_INIT_ATBC_APP)
 	test -n $(API_INIT_MSFAA_REQUEST_FOLDER)
+	test -n $(API_INIT_MSFAA_RESPONSE_FOLDER)
```
Comment (andrewsignori-aot) [inline devops/Makefile:241]: It seems that this one is missing in the env-example file.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts [hunk 4]
```diff
@@ -169,4 +176,38 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
     const repository = externalRepo ?? this.repo;
     return repository.update({ id: In(msfaaRequestIds) }, { dateRequested });
   }
+
+  /**
+   * Once the MSFAA response file is processed, updates the
+   * MSFAA received on the database with the
+   * information received. If the information was already received
+   * the record will not b updated.
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:184]: Typo on "the record will not **b** updated".

**Review by andrewsignori-aot — COMMENTED** [review]
Great work again 😉 
Please take a look at the comments. Nothing critical but I added a few comments for dates that need a closer look 👍

File: sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts [hunk 5]
```diff
@@ -51,7 +52,155 @@ export interface MSFAARequestFileLine {
...
+   * Response statuses records present on the file.
+   */
+  receivedRecords: MSFAAResponseReceivedRecord[];
+  /**
+   * Total records present on the file.
+   */
+  cancelledRecords: MSFAAResponseCancelledRecord[];
+}
+
+export class MSFAAsResponseRecordIdentification {
```
[the remark this answers is not in the document]
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts:106]: I literally added the the s to mention something like MSFAA's Response Record Identification. I can remove it, but it is better looking than the MSFAAResponseRecordIdentification IMO. :). 

I saw for SFAS and CRA we have seperate classes, I wanted to keep in a single model class as, they are more like the retrieving dummy values and doing anything in particular. We can come to a conclusion and I can change if required.

File: sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts [hunk 6]
```diff
@@ -51,7 +52,155 @@ export interface MSFAARequestFileLine {
...
+   */
+  public get newIssusingProvince(): string {
+    return this.line.substr(48, 2);
+  }
+
+  /**
+   * Date when the MSFAA was cancelled
+   */
+  public get cancelledDate(): string {
+    return this.line.substr(50, 8);
```
[the remark this answers is not in the document]
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/models/msfaa-integration.model.ts:204]: We are not using this, so did not check this conversion, but will change it.

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts [hunk 7]
```diff
@@ -180,6 +187,145 @@ export class MSFAAIntegrationService {
...
+      if (trailer.transactionCode !== TransactionCodes.MSFAATrailer) {
+        this.logger.error(
+          `The MSFAA file ${fileName} has an invalid transaction code on trailer: ${trailer.transactionCode}`,
+        );
+        // If the header is not the expected one.
+        throw new Error(
+          "The MSFAA file has an invalid transaction code on trailer",
+        );
+      }
+      // Remove trailer (not used).
```
[the remark this answers is not in the document]
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:253]: I validated before poping, so we have it in a variable

File: sources/packages/api/src/cra-integration/cra-files/cra-file-header.ts [hunk 8]
```diff
@@ -36,7 +36,7 @@ export class CRAFileHeader implements FixedFormatFileLine {
     return header.toString();
   }
 
-  public static CreateFromLine(line: string): CRAFileHeader {
+  public static createFromLine(line: string): CRAFileHeader {
```
Comment (guru-aot) [inline sources/packages/api/src/cra-integration/cra-files/cra-file-header.ts:39]: Not part of this change, but changed to be in sync

**Review by dheepak-aot — APPROVED** [review]
LGTM. Great work @guru-aot 👍

**Review by andrewsignori-aot — APPROVED** [review]
Great work 😉

**Review by ann-aot — COMMENTED** [review]
LGTM @guru-aot . Good Work 👍

**Review by ann-aot — APPROVED** [review]

**Review by abschwenker — APPROVED** [review]

**Comment by andrewsignori-aot** [comment]
Looked at the changes on the below commit about fixing the transformer and it looks good. Thanks for doing it 👍 
https://github.com/bcgov/SIMS/pull/845/commits/f91da7e88e5d185407bd5fe0732cb44c8fc6d4f9
